File: examples/device-client/csharp/ExampleScrewdriver.cs
// ─────────────────────────────────────────────────────────────────────────────
// ExampleScrewdriver — ví dụ vòng đời hoàn chỉnh dùng St4iDeviceClient (C#).
// Mô phỏng phần mềm chạy trong máy bắt vít nội bộ:
//   1) kéo recipe về máy (config-sync)  2) siết N con vít, gửi RESULT (torque+angle+waveform)
//   3) idempotent replay  4) heartbeat  5) (tùy chọn) telemetry cảm biến IoT.
//
// Chạy:  dotnet run --project examples/device-client/csharp
//   ST4I_SERVER=http://127.0.0.1:3012 ST4I_MK_KEY=mk_... [ST4I_ESP_KEY=mk_...] [ST4I_VERIFY_TLS=0]
//
// WPF: đây là console demo. Trong WPF, giữ MỘT St4iDeviceClient (singleton), gọi async,
//      chạy vòng telemetry/heartbeat trên Task nền (xem doc 61 §11.3 + St4iDeviceClient.cs header).
// ─────────────────────────────────────────────────────────────────────────────
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Threading.Tasks;
using St4i.DeviceClient;

internal static class ExampleScrewdriver
{
    private static string Env(string k, string dflt = null) =>
        Environment.GetEnvironmentVariable(k) is string v && v.Length > 0 ? v : dflt;

    private static async Task<int> Main()
    {
        string server = Env("ST4I_SERVER", "http://127.0.0.1:3012");
        string mkKey = Env("ST4I_MK_KEY");
        string espKey = Env("ST4I_ESP_KEY");
        bool verifyTls = Env("ST4I_VERIFY_TLS", "1") != "0";
        if (string.IsNullOrEmpty(mkKey)) { Console.Error.WriteLine("Thiếu ST4I_MK_KEY."); return 2; }

        // ── Máy bắt vít ──────────────────────────────────────────────────────
        using var scrw = new St4iDeviceClient(
            server, mkKey: mkKey, machineCode: "SCRW-SIM-01",
            queuePath: "scrw_queue.jsonl", verifyTls: verifyTls);

        // 1) Config-sync: kéo recipe active về máy (check → get → apply → ack)
        string cachedVersion = null;
        var recipe = new Recipe { Code = "SCRW-RC-001", Version = "1" };
        try
        {
            var sync = await scrw.SyncConfigAsync(async cfg =>
            {
                // APPLY: nạp payload recipe vào chương trình máy (ở đây chỉ in ra).
                Console.WriteLine($"  [recipe] code={cfg.Code} v{cfg.Version} payload={cfg.Payload}");
                recipe = new Recipe { Code = cfg.Code, Version = cfg.Version, Checksum = cfg.Checksum };
                await Task.CompletedTask;
                return true;
            }, configKind: "recipe", cachedVersion: cachedVersion);
            cachedVersion = sync.Version;
            Console.WriteLine($"[config-sync] changed={sync.Changed} version={sync.Version} drift={sync.DriftState}");
        }
        catch (St4iApiException e) { Console.WriteLine($"[config-sync] bỏ qua (HTTP {e.Status}): {e.Message}"); }

        // 2) Siết 5 con vít → gửi RESULT
        var rnd = new Random(42);
        for (int i = 1; i <= 5; i++)
        {
            double torque = Math.Round(11.5 + rnd.NextDouble() * 2.0, 2);   // ~11.5..13.5 Nm
            double angle = Math.Round(340 + rnd.NextDouble() * 40, 0);
            string verdict = (torque >= 10.5 && torque <= 13.5) ? "pass" : (torque > 13.5 ? "fail" : "warn");
            string unit = "DEVGUIDE-CS-" + i.ToString("D4");

            // waveform torque-vs-angle (4 điểm minh họa)
            var wave = new Waveform
            {
                Name = "torque_vs_angle", Unit = "Nm", RateHz = 1000,
                Samples = new List<double[]> { new[] { 0.0, 0.1 }, new[] { 0.1, 4.0 }, new[] { 0.2, 9.5 }, new[] { 0.3, torque } },
            };
            try
            {
                var ack = await scrw.SubmitProcessResultAsync(
                    serialNumber: unit, stepType: "screw_tightening", result: verdict,
                    recipe: recipe,
                    metrics: new[]
                    {
                        new Metric { Name = "torque", Value = torque, Unit = "Nm", Lsl = 10.5, Usl = 13.5, Nominal = 12.0 },
                        new Metric { Name = "angle",  Value = angle,  Unit = "deg" },
                    },
                    waveforms: new[] { wave },
                    idempotencyKey: $"SCRW-SIM-01:SCRW-RC-001:{i:D6}",
                    // Genealogy: stationId PHẢI là SỐ (chuỗi → 400). Field lạ top-level bị strip.
                    extra: new Dictionary<string, object> { ["lineCode"] = "LINE-01", ["lotCode"] = "LOT-2026-0001" });
                Console.WriteLine($"[{i}] {verdict.ToUpperInvariant(),-4} torque={torque}Nm angle={angle}° -> id={ack.ProcessResultId}{(ack.Duplicate ? " (dup)" : "")}");
            }
            catch (St4iNetworkException e) { Console.WriteLine($"[{i}] MẤT MẠNG -> đã xếp hàng: {e.Message}"); }
            catch (St4iApiException e) { Console.WriteLine($"[{i}] BỊ TỪ CHỐI (HTTP {e.Status} {e.Code}): {e.Message}"); }
        }

        // 3) Idempotent replay: gửi lại cùng key -> cùng id + duplicate
        var replay = await scrw.SubmitProcessResultAsync(
            serialNumber: "DEVGUIDE-CS-0001", stepType: "screw_tightening", result: "pass",
            idempotencyKey: "SCRW-SIM-01:SCRW-RC-001:000001",
            metrics: new[] { new Metric { Name = "torque", Value = 12.0, Unit = "Nm" } });
        Console.WriteLine($"[replay] id={replay.ProcessResultId} duplicate={replay.Duplicate}");

        // 4) Heartbeat
        try { var hb = await scrw.HeartbeatAsync(); Console.WriteLine($"[heartbeat] {hb}"); }
        catch (Exception e) { Console.WriteLine($"[heartbeat] {e.Message}"); }

        // 5) Telemetry cảm biến IoT (nếu có khóa ESP)
        if (!string.IsNullOrEmpty(espKey))
        {
            using var esp = new St4iDeviceClient(server, mkKey: espKey, machineCode: "ESP32-ENV-01", verifyTls: verifyTls);
            var t = await esp.SubmitTelemetryAsync(new[]
            {
                new Sample { Metric = "temperature", Value = 31.7, Unit = "C", Quality = "good" },
                new Sample { Metric = "humidity",    Value = 61.2, Unit = "%", Quality = "good" },
            });
            Console.WriteLine($"[telemetry] accepted={t.Accepted} received={t.Received}");
        }

        var flush = await scrw.FlushQueueAsync();
        Console.WriteLine($"[queue] flush cuối: sent={flush.sent} kept={flush.kept}");
        Console.WriteLine("Hoàn tất demo máy bắt vít (C#).");
        return 0;
    }
}

File: examples/device-client/python/example_screwdriver.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Ví dụ: máy BẮT VÍT gửi kết quả siết mô-men (torque) theo ST4I Process Feed v1.

Minh họa trọn vòng đời:
  1. Bootstrap khóa mk_  (claim mct_  hoặc enroll met_ — chọn 1; ở đây dùng ENV).
  2. Gửi RESULT cho từng con vít: torque + angle + waveform torque-vs-angle.
  3. Heartbeat định kỳ.
  4. Nếu mất mạng: tự xếp hàng local rồi replay ở chu trình sau.

CHẠY:
    # cách A — đã có sẵn mk_ (khuyên dùng khi máy đã onboard):
    set ST4I_SERVER=https://factory.local:5000        (Windows: set / Linux: export)
    set ST4I_MK_KEY=mk_live_xxxxxxxx
    python example_screwdriver.py

    # cách B — bootstrap lần đầu bằng claim token do admin cấp:
    set ST4I_SERVER=https://factory.local:5000
    set ST4I_CLAIM_TOKEN=mct_xxxxxxxx
    set ST4I_SERIAL=SCRW-01-DEVICE
    python example_screwdriver.py

    # cách C — zero-touch enroll (server bật ENROLLMENT_ENABLED):
    set ST4I_ENROLL_TOKEN=met_xxxxxxxx
    set ST4I_SERIAL=SCRW-01-DEVICE
    python example_screwdriver.py

Ghi chú: tất cả điểm cần điền đều đọc từ biến môi trường để KHÔNG hardcode secret
vào firmware. Với dev self-signed TLS, đặt ST4I_VERIFY_TLS=0.
"""
import os
import random
import sys
import time

# Cho phép chạy trực tiếp từ thư mục này (import client cạnh bên).
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from st4i_device_client import (  # noqa: E402
    St4iDeviceClient,
    St4iApiError,
    St4iNetworkError,
)

# ── Cấu hình (điền qua ENV — KHÔNG hardcode secret) ─────────────────────────
SERVER_URL   = os.environ.get("ST4I_SERVER", "https://localhost:5000")
MACHINE_CODE = os.environ.get("ST4I_MACHINE_CODE", "SCRW-01")   # mã máy đã đăng ký
DEVICE_SERIAL = os.environ.get("ST4I_SERIAL", "SCRW-01-DEVICE") # serial của CHÍNH máy vít
VERIFY_TLS   = os.environ.get("ST4I_VERIFY_TLS", "1") not in ("0", "false", "no")
KEY_CACHE    = os.environ.get("ST4I_KEY_FILE", "st4i_scrw01_key.json")
QUEUE_FILE   = os.environ.get("ST4I_QUEUE_FILE", "st4i_scrw01_queue.jsonl")

# Recipe + spec-limit của công đoạn siết vít M3 (self-describing — server dùng cho SPC/CPK).
RECIPE = {"code": "TQ-M3-08", "version": "2.1", "checksum": "a1b2c3d4"}
TORQUE_LSL, TORQUE_USL, TORQUE_NOMINAL = 0.70, 0.95, 0.82   # Nm
ANGLE_LSL, ANGLE_USL, ANGLE_NOMINAL = 360, 450, 410         # deg


def build_client():
    c = St4iDeviceClient(
        server_url=SERVER_URL,
        machine_code=MACHINE_CODE,
        serial_number=DEVICE_SERIAL,
        queue_path=QUEUE_FILE,       # hàng đợi BỀN qua restart
        verify_tls=VERIFY_TLS,
        max_retries=4,
    )

    # 1) Ưu tiên khóa đã cache, rồi ENV, rồi bootstrap bằng token.
    if c.load_key(KEY_CACHE):
        print(f"[key] Dùng mk_ đã lưu ({KEY_CACHE}).")
        return c
    if c.mk_key:  # từ ENV ST4I_MK_KEY
        print("[key] Dùng mk_ từ ENV ST4I_MK_KEY.")
        c.save_key(KEY_CACHE)
        return c

    claim = os.environ.get("ST4I_CLAIM_TOKEN")
    enroll = os.environ.get("ST4I_ENROLL_TOKEN")
    if claim:
        print("[bootstrap] claim mct_ -> mk_ ...")
        res = c.claim(claim, serial_number=DEVICE_SERIAL)
        print(f"[bootstrap] OK machine={res.get('code')} id={res.get('machineId')}")
    elif enroll:
        print("[bootstrap] enroll met_ -> mk_ ...")
        res = c.enroll(
            enroll,
            serial_number=DEVICE_SERIAL,
            machine_info={"name": "Screwdriver 01", "machineType": "AUTOMATION",
                          "manufacturer": "ST4I", "firmwareVersion": "1.0.0"},
        )
        print(f"[bootstrap] OK machine={res.get('code')} scopes={res.get('scopes')}")
    else:
        raise SystemExit(
            "Chưa có mk_. Đặt ST4I_MK_KEY, hoặc ST4I_CLAIM_TOKEN, hoặc ST4I_ENROLL_TOKEN."
        )
    c.save_key(KEY_CACHE)   # lưu mk_ mới nhận (chỉ show 1 lần!)
    return c


def read_torque_cycle():
    """
    Giả lập 1 chu trình siết: đọc torque/angle từ bộ điều khiển vít.
    >>> THAY BẰNG mã đọc PLC/Modbus/serial thật của bạn <<<
    Trả (torque_nm, angle_deg, waveform_samples).
    """
    torque = round(random.uniform(0.72, 0.90), 3)
    angle = random.randint(390, 430)
    # Waveform torque-vs-angle (đường cong siết) — cửa sổ quanh sự kiện, hạ mẫu để < 64KB.
    steps = [0, 90, 180, 270, 360, angle]
    samples = [[a, round(torque * (a / angle), 3)] for a in steps]
    return torque, angle, samples


def classify(torque):
    if torque < TORQUE_LSL or torque > TORQUE_USL:
        return "fail"
    if torque < TORQUE_LSL + 0.03 or torque > TORQUE_USL - 0.03:
        return "warn"   # gần biên — firmware tự phân loại (doc 57 §8.2)
    return "pass"


def main():
    c = build_client()

    # Heartbeat mở màn (xác nhận online).
    try:
        c.heartbeat()
        print("[hb] online")
    except (St4iApiError, St4iNetworkError) as e:
        print(f"[hb] cảnh báo: {e}")

    cycle = int(time.time())   # bộ đếm chu trình -> idempotencyKey ỔN ĐỊNH
    for i in range(5):         # demo 5 con vít; thực tế là vòng lặp vô hạn của máy
        cycle += 1
        unit_serial = f"SN-2026-{cycle:06d}"
        torque, angle, wave = read_torque_cycle()
        verdict = classify(torque)

        # idempotencyKey ổn định theo chu trình vật lý -> retry KHÔNG ghi trùng (§5).
        idem = f"{MACHINE_CODE}:{RECIPE['code']}:{cycle}"

        try:
            data = c.submit_process_result(
                serial_number=unit_serial,
                step_type="screw_tightening",
                result=verdict,
                recipe=RECIPE,
                metrics=[
                    {"name": "torque", "value": torque, "unit": "Nm",
                     "lsl": TORQUE_LSL, "usl": TORQUE_USL, "nominal": TORQUE_NOMINAL},
                    {"name": "angle", "value": angle, "unit": "deg",
                     "lsl": ANGLE_LSL, "usl": ANGLE_USL, "nominal": ANGLE_NOMINAL},
                ],
                waveforms=[{"name": "torque_vs_angle", "unit": "Nm", "rateHz": 500, "samples": wave}],
                idempotency_key=idem,
                # Genealogy (tùy chọn). LƯU Ý: stationId PHẢI là SỐ (platform station id) —
                # gửi chuỗi bị 400. Firmware thường chỉ biết mã trạm dạng chuỗi nên dùng
                # lineCode/lotCode (đều string). Field lạ top-level bị server STRIP âm thầm
                # (rawExtras CHƯA cài cho process-result) -> đưa số đo vendor vào metrics[].
                extra={"lineCode": "LINE-01", "lotCode": "LOT-2026-0001"},
            )
            # Replay cùng idempotencyKey trả processResultId gốc + duplicate:true (KHÔNG ghi trùng).
            pid = data.get("processResultId", "?")
            dup = " (duplicate)" if data.get("duplicate") else ""
            print(f"[{i+1}] {verdict.upper():4} torque={torque}Nm angle={angle}° -> id={pid}{dup}")
        except St4iNetworkError as e:
            # Đã tự xếp vào hàng đợi local; sẽ replay ở chu trình sau khi có mạng.
            print(f"[{i+1}] MẤT MẠNG -> đã xếp hàng: {e}")
        except St4iApiError as e:
            # 400/401/403/409: payload/khóa/version sai -> KHÔNG retry, cần sửa.
            print(f"[{i+1}] BỊ TỪ CHỐI (HTTP {e.status} {e.code}): {e}")

        time.sleep(1.0)

    # Cố xả nốt hàng đợi trước khi thoát.
    stats = c.flush_queue()
    print(f"[queue] flush cuối: sent={stats['sent']} kept={stats['kept']}")
    print("Hoàn tất demo máy bắt vít.")


if __name__ == "__main__":
    main()
